cron.php

```diff
@@ -0,0 +1,464 @@
+				call_user_func_array( 'wp_reschedule_event', $reschedule_args );
+			}
+
+			wp_unschedule_event( $event['timestamp'], $event['action'], $event['args'] );
+
+			// Run the event
+			do_action_ref_array( $event['action'], $event['args'] );
+
+			// Free process for the next event
+			if ( ! $this->is_internal_event( $event['action'] ) ) {
+				$this->free_lock();
+			}
+
+			$time_end = microtime( true );
+
+			return rest_ensure_response( array(
+				'success' => true,
+				'message' => sprintf( __( 'Job with action `%1$s` and arguments `%2$s` completed in %3$d seconds.', 'wp-cron-control-revisited' ), $event['action'], serialize( $event['args'] ), $time_end - $time_start ),
+			) );
+		} else {
+			return new WP_Error( 'no-event', __( 'The specified event could not be found.', 'wp-cron-control-revisited' ) );
+		}
+	}
+
+	/**
+	 * Find an event's data using its hashed representations
+	 */
+	private function get_event( $timestamp, $action_hashed, $instance ) {
+		$events = get_option( 'cron' );
+		$event  = false;
+
+		if ( isset( $events[ $timestamp ] ) ) {
+			foreach ( $events[ $timestamp ] as $action => $action_events ) {
+				if ( hash_equals( md5( $action ), $action_hashed ) && isset( $action_events[ $instance ] ) ) {
+					$event              = $action_events[ $instance ];
+					$event['timestamp'] = $timestamp;
+					$event['action']    = $action;
+					break;
+				}
+			}
+		}
+
+		return $event;
+	}
+
+	/**
+	 * PLUGIN UTILITY METHODS
+	 */
+
+	/**
+	 * Check if request is authorized
+	 */
+	public function check_secret( $request ) {
+		$body = $request->get_json_params();
+
+		// For now, mimic original plugin's "authentication" method. This needs to be better.
+		if ( ! isset( $body['secret'] ) || ! hash_equals( $this->secret, $body['secret'] ) ) {
+			return new WP_Error( 'no-secret', __( 'Secret must be specified with all requests', 'wp-cron-control-revisited' ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Events that are always run, regardless of how many jobs are queued
+	 */
+	private function is_internal_event( $action ) {
+		return in_array( $action, wp_list_pluck( $this->internal_jobs, 'action' ) );
+	}
+
+	/**
+	 * Allow specific events to be blocked perpetually
+	 */
+	private function is_blocked_event( $action ) {
+		$blocked_hooks = array();
+
+		return in_array( $action, $blocked_hooks );
+	}
+
+	/**
+	 * Set a lock and limit how many concurrent jobs are permitted
+	 */
+	private function check_lock() {
+		// Prevent deadlock
+		if ( (int) wp_cache_get( $this->cache_key_lock_timestamp, null, true ) < time() - $this->job_timeout_in_minutes * MINUTE_IN_SECONDS ) {
+			wp_cache_set( $this->cache_key_lock, 0 );
+			wp_cache_set( $this->cache_key_lock_timestamp, time() );
+			return true;
+		}
+
+		// Check if process can run
+		if ( (int) wp_cache_get( $this->cache_key_lock, null, true ) >= $this->job_concurrency_limit ) {
+			return false;
+		} else {
```

Comment:
I know, there's just something about the symmetry and explicitness that I'm fond of.